# Tiền xử lý dữ liệu

## Đọc thư viện cần thiết

In [1]:
import numpy as np
import pandas as pd

In [2]:
# Đọc dữ liệu từ pandas
df = pd.read_csv('../Dataset/archive/WA_Fn-UseC_-HR-Employee-Attrition.csv')

## Xử lý dữ liệu

Dữ liệu sau khi đọc

In [3]:
df

Age Attrition     BusinessTravel  DailyRate              Department  \
0      41       Yes      Travel_Rarely       1102                   Sales   
1      49        No  Travel_Frequently        279  Research & Development   
2      37       Yes      Travel_Rarely       1373  Research & Development   
3      33        No  Travel_Frequently       1392  Research & Development   
4      27        No      Travel_Rarely        591  Research & Development   
...   ...       ...                ...        ...                     ...   
1465   36        No  Travel_Frequently        884  Research & Development   
1466   39        No      Travel_Rarely        613  Research & Development   
1467   27        No      Travel_Rarely        155  Research & Development   
1468   49        No  Travel_Frequently       1023                   Sales   
1469   34        No      Travel_Rarely        628  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  \
0                    1          2  Life Sciences              1   
1                    8          1  Life Sciences              1   
2                    2          2          Other              1   
3                    3          4  Life Sciences              1   
4                    2          1        Medical              1   
...                ...        ...            ...            ...   
1465                23          2        Medical              1   
1466                 6          1        Medical              1   
1467                 4          3  Life Sciences              1   
1468                 2          3        Medical              1   
1469                 8          3        Medical              1   

      EmployeeNumber  ...  RelationshipSatisfaction StandardHours  \
0                  1  ...                         1            80   
1                  2  ...                         4            80   
2                  4  ...                         2            80   
3                  5  ...                         3            80   
4                  7  ...                         4            80   
...              ...  ...                       ...           ...   
1465            2061  ...                         3            80   
1466            2062  ...                         1            80   
1467            2064  ...                         2            80   
1468            2065  ...                         4            80   
1469            2068  ...                         1            80   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                    0                  8                      0   
1                    1                 10                      3   
2                    0                  7                      3   
3                    0                  8                      3   
4                    1                  6                      3   
...                ...                ...                    ...   
1465                 1                 17                      3   
1466                 1                  9                      5   
1467                 1                  6                      0   
1468                 0                 17                      3   
1469                 0                  6                      3   

     WorkLifeBalance  YearsAtCompany YearsInCurrentRole  \
0                  1               6                  4   
1                  3              10                  7   
2                  3               0                  0   
3                  3               8                  7   
4                  3               2                  2   
...              ...             ...                ...   
1465               3               5                  2   
1466               3               7                  7   
1467               3               6                  2   
1468               2               9    

Mô tả sơ bộ về dữ liệu của tập dataset trên

In [4]:
df.describe()

Age    DailyRate  DistanceFromHome    Education  EmployeeCount  \
count  1470.000000  1470.000000       1470.000000  1470.000000         1470.0   
mean     36.923810   802.485714          9.192517     2.912925            1.0   
std       9.135373   403.509100          8.106864     1.024165            0.0   
min      18.000000   102.000000          1.000000     1.000000            1.0   
25%      30.000000   465.000000          2.000000     2.000000            1.0   
50%      36.000000   802.000000          7.000000     3.000000            1.0   
75%      43.000000  1157.000000         14.000000     4.000000            1.0   
max      60.000000  1499.000000         29.000000     5.000000            1.0   

       EmployeeNumber  EnvironmentSatisfaction   HourlyRate  JobInvolvement  \
count     1470.000000              1470.000000  1470.000000     1470.000000   
mean      1024.865306                 2.721769    65.891156        2.729932   
std        602.024335                 1.093082    20.329428        0.711561   
min          1.000000                 1.000000    30.000000        1.000000   
25%        491.250000                 2.000000    48.000000        2.000000   
50%       1020.500000                 3.000000    66.000000        3.000000   
75%       1555.750000                 4.000000    83.750000        3.000000   
max       2068.000000                 4.000000   100.000000        4.000000   

          JobLevel  ...  RelationshipSatisfaction  StandardHours  \
count  1470.000000  ...               1470.000000         1470.0   
mean      2.063946  ...                  2.712245           80.0   
std       1.106940  ...                  1.081209            0.0   
min       1.000000  ...                  1.000000           80.0   
25%       1.000000  ...                  2.000000           80.0   
50%       2.000000  ...                  3.000000           80.0   
75%       3.000000  ...                  4.000000           80.0   
max       5.000000  ...                  4.000000           80.0   

       StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
count       1470.000000        1470.000000            1470.000000   
mean           0.793878          11.279592               2.799320   
std            0.852077           7.780782               1.289271   
min            0.000000           0.000000               0.000000   
25%            0.000000           6.000000               2.000000   
50%            1.000000          10.000000               3.000000   
75%            1.000000          15.000000               3.000000   
max            3.000000          40.000000               6.000000   

       WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
count      1470.000000     1470.000000         1470.000000   
mean          2.761224        7.008163            4.229252   
std           0.706476        6.126525            3.623137   
min           1.000000        0.000000            0.000000   
25%           2.000000        3.000000            2.000000   
50%           3.000000        5.000000            3.000000   
75%           3.000000        9.000000            7.000000   
max           4.000000       40.000000           18.000000   

       YearsSinceLastPromotion  YearsWithCurrManager  
count              1470.000000           1470.000000  
mean                  2.187755              4.123129  
std                   3.222430              3.568136  
min                   0.000000              0.000000  
25%                   0.000000              2.000000  
50%                   1.000000              3.000000  
75%                   3.000000              7.000000  
max                  15.000000             17.000000  

[8 rows x 26 columns]

In [5]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

Nhận xét:

Dữ liệu ban đầu có tổng cộng 35 cột trong đó có:
- 01 cột nhãn quyết định xem có nghỉ việc hay không.
- 34 cột còn lại là các trường của dữ liệu.
- Các trường dữ liệu của tập dữ liệu chứa cả dữ liệu liên tục và dữ liệu rời rạc, không chứa dữ liệu Null hoặc các dữ liệu đặc biệt.

Tách lấy các cột X (các trường dữ liệu), cột y (nhãn của dữ liệu).

In [6]:
X = df.drop(['Attrition', 'EmployeeNumber'], axis=1) # tách thêm Empl number: đó là ID của các nhân viên
y = df['Attrition']

In [7]:
print(f'Số bản ghi dữ liệu: {len(X)}')
print(f'Số cột dữ liệu (tập X): {len(X.columns)}')


Số bản ghi dữ liệu: 1470
Số cột dữ liệu (tập X): 33


Thống kê chi tiết các trường.

In [8]:
columns = X.columns

for col in columns:
    print(f'Thống kê cho cột {col}')
    print(X[col].value_counts())
    print('=' * 50)


Thống kê cho cột Age
Age
35    78
34    77
36    69
31    69
29    68
32    61
30    60
38    58
33    58
40    57
37    50
28    48
27    48
42    46
39    42
45    41
41    40
26    39
44    33
46    33
43    32
50    30
25    26
24    26
49    24
47    24
55    22
48    19
51    19
53    19
54    18
52    18
22    16
56    14
58    14
23    14
21    13
20    11
59    10
19     9
18     8
60     5
57     4
Name: count, dtype: int64
Thống kê cho cột BusinessTravel
BusinessTravel
Travel_Rarely        1043
Travel_Frequently     277
Non-Travel            150
Name: count, dtype: int64
Thống kê cho cột DailyRate
DailyRate
691     6
1082    5
329     5
530     5
408     5
       ..
1382    1
161     1
801     1
642     1
1154    1
Name: count, Length: 886, dtype: int64
Thống kê cho cột Department
Department
Research & Development    961
Sales                     446
Human Resources            63
Name: count, dtype: int64
Thống kê cho cột DistanceFromHome
DistanceFromHome
2     211
1     208

X ban đầu:

In [9]:
X

Age     BusinessTravel  DailyRate              Department  \
0      41      Travel_Rarely       1102                   Sales   
1      49  Travel_Frequently        279  Research & Development   
2      37      Travel_Rarely       1373  Research & Development   
3      33  Travel_Frequently       1392  Research & Development   
4      27      Travel_Rarely        591  Research & Development   
...   ...                ...        ...                     ...   
1465   36  Travel_Frequently        884  Research & Development   
1466   39      Travel_Rarely        613  Research & Development   
1467   27      Travel_Rarely        155  Research & Development   
1468   49  Travel_Frequently       1023                   Sales   
1469   34      Travel_Rarely        628  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  \
0                    1          2  Life Sciences              1   
1                    8          1  Life Sciences              1   
2                    2          2          Other              1   
3                    3          4  Life Sciences              1   
4                    2          1        Medical              1   
...                ...        ...            ...            ...   
1465                23          2        Medical              1   
1466                 6          1        Medical              1   
1467                 4          3  Life Sciences              1   
1468                 2          3        Medical              1   
1469                 8          3        Medical              1   

      EnvironmentSatisfaction  Gender  ...  RelationshipSatisfaction  \
0                           2  Female  ...                         1   
1                           3    Male  ...                         4   
2                           4    Male  ...                         2   
3                           4  Female  ...                         3   
4                           1    Male  ...                         4   
...                       ...     ...  ...                       ...   
1465                        3    Male  ...                         3   
1466                        4    Male  ...                         1   
1467                        2    Male  ...                         2   
1468                        4    Male  ...                         4   
1469                        2    Male  ...                         1   

      StandardHours  StockOptionLevel TotalWorkingYears  \
0                80                 0                 8   
1                80                 1                10   
2                80                 0                 7   
3                80                 0                 8   
4                80                 1                 6   
...             ...               ...               ...   
1465             80                 1                17   
1466             80                 1                 9   
1467             80                 1                 6   
1468             80                 0                17   
1469             80                 0                 6   

      TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                         0               1               6   
1                         3               3              10   
2                         3               3               0   
3                         3               3               8   
4                         3               3               2   
...                     ...             ...             ...   
1465                      3               3               5   
1466                      5               3               7   
1467                      0               3               6   
1468                      3               2               9   
1469                      3               4               4   

      YearsInCurrentRole  YearsSinceLastPromotion YearsWithCurrMa

y ban đầu:

In [10]:
y

0       Yes
1        No
2       Yes
3        No
4        No
       ... 
1465     No
1466     No
1467     No
1468     No
1469     No
Name: Attrition, Length: 1470, dtype: object

Về độ cân bằng của dữ liệu.

In [11]:
count_yes = np.sum(y == 'Yes')
count_no = np.sum(y == 'No')

print(f'Count yes: {count_yes}')
print(f'Count no: {count_no}')
print(f'yes / no: {float(count_yes) / float(count_yes + count_no):.2f}')

Count yes: 237
Count no: 1233
yes / no: 0.16


## Mã hóa dữ liệu ban đầu bằng Onehot encode trong thư viện sklearn

In [12]:
from sklearn.preprocessing import OneHotEncoder, LabelEncoder

# sử dụng OHE để mã hóa các categories của dữ liệu, Label Encode để mã hóa lại đầu ra của y
one_hot_encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore')
label_encoder = LabelEncoder()


In [13]:
X_origin = X.copy()
y_origin = y.copy()

In [14]:
X_origin.shape

(1470, 33)

In [15]:
from sklearn.preprocessing import OneHotEncoder
import pandas as pd

# Giả sử X_origin là DataFrame ban đầu
encode_cols = [
    'BusinessTravel', 'Department', 'Education', 'EducationField',
    'EnvironmentSatisfaction', 'Gender', 'JobInvolvement', 'JobLevel',
    'JobRole', 'JobSatisfaction', 'MaritalStatus', 'Over18', 'OverTime',
    'PerformanceRating', 'RelationshipSatisfaction', 'StockOptionLevel',
    'WorkLifeBalance'
]
one_hot_encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore')
# fit chuyển đổi luôn
X_encoded_array = one_hot_encoder.fit_transform(X_origin[encode_cols]) 

# Lấy tên cột mới sau khi mã hóa
encoded_columns = one_hot_encoder.get_feature_names_out(encode_cols)

# Tạo DataFrame từ mảng mã hóa
X_encoded_df = pd.DataFrame(X_encoded_array, columns=encoded_columns, index=X_origin.index)

# Gộp với phần dữ liệu còn lại
X_encode = pd.concat([X_origin.drop(columns=encode_cols), X_encoded_df], axis=1)

In [16]:
X_encode.shape

(1470, 81)

In [17]:
X_encode

Age  DailyRate  DistanceFromHome  EmployeeCount  HourlyRate  \
0      41       1102                 1              1          94   
1      49        279                 8              1          61   
2      37       1373                 2              1          92   
3      33       1392                 3              1          56   
4      27        591                 2              1          40   
...   ...        ...               ...            ...         ...   
1465   36        884                23              1          41   
1466   39        613                 6              1          42   
1467   27        155                 4              1          87   
1468   49       1023                 2              1          63   
1469   34        628                 8              1          82   

      MonthlyIncome  MonthlyRate  NumCompaniesWorked  PercentSalaryHike  \
0              5993        19479                   8                 11   
1              5130        24907                   1                 23   
2              2090         2396                   6                 15   
3              2909        23159                   1                 11   
4              3468        16632                   9                 12   
...             ...          ...                 ...                ...   
1465           2571        12290                   4                 17   
1466           9991        21457                   4                 15   
1467           6142         5174                   1                 20   
1468           5390        13243                   2                 14   
1469           4404        10228                   2                 12   

      StandardHours  ...  RelationshipSatisfaction_3  \
0                80  ...                         0.0   
1                80  ...                         0.0   
2                80  ...                         0.0   
3                80  ...                         1.0   
4                80  ...                         0.0   
...             ...  ...                         ...   
1465             80  ...                         1.0   
1466             80  ...                         0.0   
1467             80  ...                         0.0   
1468             80  ...                         0.0   
1469             80  ...                         0.0   

      RelationshipSatisfaction_4  StockOptionLevel_0  StockOptionLevel_1  \
0                            0.0                 1.0                 0.0   
1                            1.0                 0.0                 1.0   
2                            0.0                 1.0                 0.0   
3                            0.0                 1.0                 0.0   
4                            1.0                 0.0                 1.0   
...                          ...                 ...                 ...   
1465                         0.0                 0.0                 1.0   
1466                         0.0                 0.0                 1.0   
1467                         0.0                 0.0                 1.0   
1468                         1.0                 1.0                 0.0   
1469                         0.0                 1.0                 0.0   

      StockOptionLevel_2  StockOptionLevel_3  WorkLifeBalance_1  \
0                    0.0                 0.0                1.0   
1                    0.0                 0.0                0.0   
2                    0.0                 0.0                0.0   
3                    0.0                 0.0                0.0   
4                    0.0                 0.0                0.0   
...                  ...                 ...                ...   
1465                 0.0                 0.0                0.0   
1466                 0.0                 0.0                0.0   
1467                 0.0                 0.0                0.0   
1468                

In [18]:
X_encode.to_csv("/home/trong-thanh/Downloads/encode.csv", index=False) # ghi ra file để test

Nhận xét: 

Dữ liệu X ban đầu code shape là: (1470, 33)
Sau khi encode thì X có dạng: (1470, 81)
Các cột sau khi qua One hot encode thì các categories sẽ tạo thêm cột mới trương ứng với các categories để biểu diễn dữ liệu. Ví dụ cột A chứa kiểu cate như: A1, A2, A3 thì sau khi encode thì sẽ tạo thêm các cột mới và đánh dẫu xem mẫu nào thì tương ứng với A1, 2, 3...

Mã hóa Labels về dạng số sử dụng Labels Encode

In [19]:
y_encode = label_encoder.fit_transform(y)

In [20]:
y_encode

array([1, 0, 1, ..., 0, 0, 0], shape=(1470,))

Nhận xét: 
Dữ liệu ban đầu chỉ có `Yes` và `No` thì sau khi mã hóa thì đã chuyển về tương ứng với `1` và `0` để dễ dàng trong việc tính toán.

## Chia dữ liệu

In [21]:
X_origin = X.copy()
y_origin = y.copy()

In [22]:
continous_data = [
    'Age',
    'DailyRate',
    'DistanceFromHome',
    'EmployeeCount',
    'HourlyRate',
    'MonthlyIncome',
    'MonthlyRate',
    'NumCompaniesWorked',
    'PercentSalaryHike',
    'StandardHours',
    'TotalWorkingYears',
    'TrainingTimesLastYear',
    'YearsAtCompany',
    'YearsInCurrentRole',
    'YearsSinceLastPromotion',
    'YearsWithCurrManager'
]

classified_data = [
    'BusinessTravel',
    'Department',
    'Education',
    'EducationField',
    'EnvironmentSatisfaction',
    'Gender',
    'JobInvolvement',
    'JobLevel',
    'JobRole',
    'JobSatisfaction',
    'MaritalStatus',
    'Over18',
    'OverTime',
    'PerformanceRating',
    'RelationshipSatisfaction',
    'StockOptionLevel',
    'WorkLifeBalance'
]

X_continous = X_origin[continous_data]
X_classifieds = X_origin[classified_data]

### Dữ liệu liên tục

In [23]:
X_continous

Age  DailyRate  DistanceFromHome  EmployeeCount  HourlyRate  \
0      41       1102                 1              1          94   
1      49        279                 8              1          61   
2      37       1373                 2              1          92   
3      33       1392                 3              1          56   
4      27        591                 2              1          40   
...   ...        ...               ...            ...         ...   
1465   36        884                23              1          41   
1466   39        613                 6              1          42   
1467   27        155                 4              1          87   
1468   49       1023                 2              1          63   
1469   34        628                 8              1          82   

      MonthlyIncome  MonthlyRate  NumCompaniesWorked  PercentSalaryHike  \
0              5993        19479                   8                 11   
1              5130        24907                   1                 23   
2              2090         2396                   6                 15   
3              2909        23159                   1                 11   
4              3468        16632                   9                 12   
...             ...          ...                 ...                ...   
1465           2571        12290                   4                 17   
1466           9991        21457                   4                 15   
1467           6142         5174                   1                 20   
1468           5390        13243                   2                 14   
1469           4404        10228                   2                 12   

      StandardHours  TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                80                  8                      0               6   
1                80                 10                      3              10   
2                80                  7                      3               0   
3                80                  8                      3               8   
4                80                  6                      3               2   
...             ...                ...                    ...             ...   
1465             80                 17                      3               5   
1466             80                  9                      5               7   
1467             80                  6                      0               6   
1468             80                 17                      3               9   
1469             80                  6                      3               4   

      YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                      4                        0                     5  
1                      7                        1                     7  
2                      0                        0                     0  
3                      7                        3                     0  
4                      2                        2                     2  
...                  ...                      ...                   ...  
1465                   2                        0                     3  
1466                   7                        1                     7  
1467                   2                        0                     3  
1468                   6                        0                     8  
1469                   3                        1                     2  

[1470 rows x 16 columns]

### Dữ liệu rời rạc

In [24]:
X_classifieds

BusinessTravel              Department  Education EducationField  \
0         Travel_Rarely                   Sales          2  Life Sciences   
1     Travel_Frequently  Research & Development          1  Life Sciences   
2         Travel_Rarely  Research & Development          2          Other   
3     Travel_Frequently  Research & Development          4  Life Sciences   
4         Travel_Rarely  Research & Development          1        Medical   
...                 ...                     ...        ...            ...   
1465  Travel_Frequently  Research & Development          2        Medical   
1466      Travel_Rarely  Research & Development          1        Medical   
1467      Travel_Rarely  Research & Development          3  Life Sciences   
1468  Travel_Frequently                   Sales          3        Medical   
1469      Travel_Rarely  Research & Development          3        Medical   

      EnvironmentSatisfaction  Gender  JobInvolvement  JobLevel  \
0                           2  Female               3         2   
1                           3    Male               2         2   
2                           4    Male               2         1   
3                           4  Female               3         1   
4                           1    Male               3         1   
...                       ...     ...             ...       ...   
1465                        3    Male               4         2   
1466                        4    Male               2         3   
1467                        2    Male               4         2   
1468                        4    Male               2         2   
1469                        2    Male               4         2   

                        JobRole  JobSatisfaction MaritalStatus Over18  \
0               Sales Executive                4        Single      Y   
1            Research Scientist                2       Married      Y   
2         Laboratory Technician                3        Single      Y   
3            Research Scientist                3       Married      Y   
4         Laboratory Technician                2       Married      Y   
...                         ...              ...           ...    ...   
1465      Laboratory Technician                4       Married      Y   
1466  Healthcare Representative                1       Married      Y   
1467     Manufacturing Director                2       Married      Y   
1468            Sales Executive                2       Married      Y   
1469      Laboratory Technician                3       Married      Y   

     OverTime  PerformanceRating  RelationshipSatisfaction  StockOptionLevel  \
0         Yes                  3                         1                 0   
1          No                  4                         4                 1   
2         Yes                  3                         2                 0   
3         Yes                  3                         3                 0   
4          No                  3                         4                 1   
...       ...                ...                       ...               ...   
1465       No                  3                         3                 1   
1466       No                  3                         1                 1   
1467      Yes                  4                         2                 1   
1468       No                  3                         4                 0   
1469       No                  3                         1                 0   

      WorkLifeBalance  
0                   1  
1                   3  
2                   3  
3                   3  
4                   3  
...               ...  
1465                3  
1466                3  
1467                3  
1468                2  
1469                4  

[1470 rows x 17 columns]

### Chuẩn hóa dữ liệu liên tục

In [25]:
X_continous_std = X_continous.copy()

In [26]:
X_continous_std

Age  DailyRate  DistanceFromHome  EmployeeCount  HourlyRate  \
0      41       1102                 1              1          94   
1      49        279                 8              1          61   
2      37       1373                 2              1          92   
3      33       1392                 3              1          56   
4      27        591                 2              1          40   
...   ...        ...               ...            ...         ...   
1465   36        884                23              1          41   
1466   39        613                 6              1          42   
1467   27        155                 4              1          87   
1468   49       1023                 2              1          63   
1469   34        628                 8              1          82   

      MonthlyIncome  MonthlyRate  NumCompaniesWorked  PercentSalaryHike  \
0              5993        19479                   8                 11   
1              5130        24907                   1                 23   
2              2090         2396                   6                 15   
3              2909        23159                   1                 11   
4              3468        16632                   9                 12   
...             ...          ...                 ...                ...   
1465           2571        12290                   4                 17   
1466           9991        21457                   4                 15   
1467           6142         5174                   1                 20   
1468           5390        13243                   2                 14   
1469           4404        10228                   2                 12   

      StandardHours  TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                80                  8                      0               6   
1                80                 10                      3              10   
2                80                  7                      3               0   
3                80                  8                      3               8   
4                80                  6                      3               2   
...             ...                ...                    ...             ...   
1465             80                 17                      3               5   
1466             80                  9                      5               7   
1467             80                  6                      0               6   
1468             80                 17                      3               9   
1469             80                  6                      3               4   

      YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                      4                        0                     5  
1                      7                        1                     7  
2                      0                        0                     0  
3                      7                        3                     0  
4                      2                        2                     2  
...                  ...                      ...                   ...  
1465                   2                        0                     3  
1466                   7                        1                     7  
1467                   2                        0                     3  
1468                   6                        0                     8  
1469                   3                        1                     2  

[1470 rows x 16 columns]

Để chuẩn hóa dữ liệu, ở đây chúng tôi sử dụng gói `sklearn.preprocessing.StandardScaler` trong thư viện sklearn.

Về  `StandardScaler` là một kỹ thuật chia tỷ lệ đặc trưng tuân theo phân phối chuẩn và được sử dụng để chuẩn hóa giá trị của các đặc trưng số. Nó biến đổi dữ liệu sao cho giá trị trung bình bằng 0 và độ lệch chuẩn bằng 1. Kỹ thuật này lý tưởng cho các thuật toán như SVM, hồi quy logistic hoặc mạng nơ-ron nhân tạo giả định dữ liệu phân phối chuẩn.

Phương pháp này trừ đi giá trị trung bình của dữ liệu và chia cho độ lệch chuẩn. Phương pháp này tập trung dữ liệu quanh giá trị 0 và chuẩn hóa độ biến thiên:

$$
X_{\text{scaled}} = \frac{X - \mu}{\sigma}
$$

Trong đó:
- `X` là giá trị ban đầu,
- `μ` là giá trị trung bình của tính năng,
- `σ` là độ lệch chuẩn.

Về ưu điểm:
- Nó xử lý các tính năng với các đơn vị khác nhau một cách hiệu quả.
- Nó làm giảm tác động của các giá trị ngoại lai mà không loại bỏ chúng hoàn toàn.

Về nhược điểm:
- Nó nhạy cảm với các giá trị ngoại lệ và các giá trị cực đoan có thể làm lệch giá trị trung bình và độ lệch chuẩn, dẫn đến khả năng mở rộng kém.
- Nó không lý tưởng cho các phân phối không chuẩn.


Thực hiện chuẩn hóa.

In [27]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_continous_std = scaler.fit_transform(X_continous_std)

In [28]:
X_continous_std

array([[ 0.4463504 ,  0.74252653, -1.01090934, ..., -0.0632959 ,
        -0.67914568,  0.24583399],
       [ 1.32236521, -1.2977746 , -0.14714972, ...,  0.76499762,
        -0.36871529,  0.80654148],
       [ 0.008343  ,  1.41436324, -0.88751511, ..., -1.16768726,
        -0.67914568, -1.15593471],
       ...,
       [-1.08667552, -1.60518328, -0.64072665, ..., -0.61549158,
        -0.67914568, -0.31487349],
       [ 1.32236521,  0.54667746, -0.88751511, ...,  0.48889978,
        -0.67914568,  1.08689522],
       [-0.32016256, -0.43256792, -0.14714972, ..., -0.33939374,
        -0.36871529, -0.59522723]], shape=(1470, 16))

In [29]:
# đưa về dạng df cho dễ nhìn
X_continous_std_df = pd.DataFrame(X_continous_std)

In [30]:
X_continous_std_df

0         1         2    3         4         5         6   \
0     0.446350  0.742527 -1.010909  0.0  1.383138 -0.108350  0.726020   
1     1.322365 -1.297775 -0.147150  0.0 -0.240677 -0.291719  1.488876   
2     0.008343  1.414363 -0.887515  0.0  1.284725 -0.937654 -1.674841   
3    -0.429664  1.461466 -0.764121  0.0 -0.486709 -0.763634  1.243211   
4    -1.086676 -0.524295 -0.887515  0.0 -1.274014 -0.644858  0.325900   
...        ...       ...       ...  ...       ...       ...       ...   
1465 -0.101159  0.202082  1.703764  0.0 -1.224807 -0.835451 -0.284329   
1466  0.227347 -0.469754 -0.393938  0.0 -1.175601  0.741140  1.004010   
1467 -1.086676 -1.605183 -0.640727  0.0  1.038693 -0.076690 -1.284418   
1468  1.322365  0.546677 -0.887515  0.0 -0.142264 -0.236474 -0.150393   
1469 -0.320163 -0.432568 -0.147150  0.0  0.792660 -0.445978 -0.574124   

            7         8    9         10        11        12        13  \
0     2.125136 -1.150554  0.0 -0.421642 -2.171982 -0.164613 -0.063296   
1    -0.678049  2.129306  0.0 -0.164511  0.155707  0.488508  0.764998   
2     1.324226 -0.057267  0.0 -0.550208  0.155707 -1.144294 -1.167687   
3    -0.678049 -1.150554  0.0 -0.421642  0.155707  0.161947  0.764998   
4     2.525591 -0.877232  0.0 -0.678774  0.155707 -0.817734 -0.615492   
...        ...       ...  ...       ...       ...       ...       ...   
1465  0.523316  0.489376  0.0  0.735447  0.155707 -0.327893 -0.615492   
1466  0.523316 -0.057267  0.0 -0.293077  1.707500 -0.001333  0.764998   
1467 -0.678049  1.309341  0.0 -0.678774 -2.171982 -0.164613 -0.615492   
1468 -0.277594 -0.330589  0.0  0.735447  0.155707  0.325228  0.488900   
1469 -0.277594 -0.877232  0.0 -0.678774  0.155707 -0.491174 -0.339394   

            14        15  
0    -0.679146  0.245834  
1    -0.368715  0.806541  
2    -0.679146 -1.155935  
3     0.252146 -1.155935  
4    -0.058285 -0.595227  
...        ...       ...  
1465 -0.679146 -0.314873  
1466 -0.368715  0.806541  
1467 -0.679146 -0.314873  
1468 -0.679146  1.086895  
1469 -0.368715 -0.595227  

[1470 rows x 16 columns]

#### Mô tả dữ liệu sau khi chuẩn hóa

Trước khi chuẩn hóa, dữ liệu của các trường `Age`, `DailyRate`, .. có độ lớn rất khác nhau, điều này gây mất cân bằng thang đo cho các thuộc tính, các biến giá trị lớn như `MonthlyIncome` hay `MonthlyRate` có thể sẽ chi phối mô hình.

Sau khi chuẩn hóa bằng `StandardScaler`, Các giá trị dao động quanh 0, thường nằm trong khoảng từ -3 đến 3, Các cột có trung bình ≈ 0, độ lệch chuẩn ≈ 1

In [31]:
X_continous_std_df.describe()

0             1             2       3             4   \
count  1.470000e+03  1.470000e+03  1.470000e+03  1470.0  1.470000e+03   
mean  -3.504377e-17  5.075305e-17  4.350262e-17     0.0  1.691768e-16   
std    1.000340e+00  1.000340e+00  1.000340e+00     0.0  1.000340e+00   
min   -2.072192e+00 -1.736576e+00 -1.010909e+00     0.0 -1.766079e+00   
25%   -7.581700e-01 -8.366616e-01 -8.875151e-01     0.0 -8.803615e-01   
50%   -1.011589e-01 -1.204135e-03 -2.705440e-01     0.0  5.355811e-03   
75%    6.653541e-01  8.788772e-01  5.932157e-01     0.0  8.787715e-01   
max    2.526886e+00  1.726730e+00  2.444129e+00     0.0  1.678377e+00   

                 5             6             7             8       9   \
count  1.470000e+03  1.470000e+03  1.470000e+03  1.470000e+03  1470.0   
mean  -4.471102e-17  3.021015e-17  1.450087e-17  2.271803e-16     0.0   
std    1.000340e+00  1.000340e+00  1.000340e+00  1.000340e+00     0.0   
min   -1.167343e+00 -1.717284e+00 -1.078504e+00 -1.150554e+00     0.0   
25%   -7.632087e-01 -8.806440e-01 -6.780494e-01 -8.772324e-01     0.0   
50%   -3.365516e-01 -1.090645e-02 -2.775943e-01 -3.305891e-01     0.0   
75%    3.986245e-01  8.641014e-01  5.233157e-01  7.626976e-01     0.0   
max    2.867626e+00  1.782888e+00  2.525591e+00  2.675949e+00     0.0   

                 10            11            12            13            14  \
count  1.470000e+03  1.470000e+03  1.470000e+03  1.470000e+03  1.470000e+03   
mean  -1.208406e-18  8.700523e-17 -3.021015e-17  9.063045e-17  1.208406e-18   
std    1.000340e+00  1.000340e+00  1.000340e+00  1.000340e+00  1.000340e+00   
min   -1.450167e+00 -2.171982e+00 -1.144294e+00 -1.167687e+00 -6.791457e-01   
25%   -6.787735e-01 -6.201892e-01 -6.544537e-01 -6.154916e-01 -6.791457e-01   
50%   -1.645114e-01  1.557071e-01 -3.278933e-01 -3.393937e-01 -3.687153e-01   
75%    4.783162e-01  1.557071e-01  3.252275e-01  7.649976e-01  2.521455e-01   
max    3.692454e+00  2.483396e+00  5.386914e+00  3.802074e+00  3.977310e+00   

                 15  
count  1.470000e+03  
mean  -2.779334e-17  
std    1.000340e+00  
min   -1.155935e+00  
25%   -5.952272e-01  
50%   -3.148735e-01  
75%    8.065415e-01  
max    3.610079e+00

Trên đây là file thống kê chi tiết.